File: emda/boundary.py
class boundary:
    """
    boundary conditions
    """
    pass


class apply(boundary):
    """
    apply some boundary condition
    """


    def __init__(self, box_size):
        """
        Parameters
        ----------
        box_size : numpy array of three floats
            box size in x, y, z
        """
        self.box_size = box_size


    def pbc(self, natoms, positions):
        """
        applies periodic boundary conditions to the particles

        Parameters
        ----------
        natoms : integer
            number of atoms

        positions : numpy array with float32 data
            unwrapped positions in the SoA convention

        Returns
        -------
        positions : numpy array with float32 dat
            wrapped positions
        """
        for i in range(0,3):
            for j in range(0, natoms):
                while (positions[i*natoms + j] < 0.0):
                    positions[i*natoms + j] += self.box_size[i]
                while (positions[i*natoms + j] > self.box_size[i]):
                    positions[i*natoms + j] -= self.box_size[i]

        return positions


    def minimum_image(self, x_central, x_interact):
        """
        nearest particle in minimum image convention

        Parameters
        ----------
        x_central : numpy array with three floats
            coordinates of the central atom

        x_interact : numpy array with three floats
            coordinates of the interact atom

        Returns
        -------
        x_ci : numpy array with three float32
            components of the vector that separates the atom interact from the
            atom central
        """
        x_ci = np.zeros(3, dtype=np.float32)

        for i in range(0,3):
            x_ci[i] = x_central[i] - x_interact[i]
            while (x_ci[i] > 0.5 * self.box_size[i]): x_ci[i] -= self.box_size[i]
            while (x_ci[i] < -0.5 * self.box_size[i]): x_ci[i] += self.box_size[i]
        
        return x_ci

File: exma/boundary/boundary.c
#include "boundary.h"
#include <stdio.h>

void pbc(const int N, const float *box_size, float *x){
    /* periodic boundary conditions */
    for (int j = 0; j < 3; j++){
        for (int i = 0; i < N; i++){
            while (x[j*N + i] < 0.0) x[j*N + i] += box_size[j];
            while (x[j*N + i] > box_size[j]) x[j*N + i] -= box_size[j];
        }
    }
}


void minimum_image(const float *box_size, const float *central,
                   const float *interact, float *x_ci){
    /* minimum image */
    for (int i = 0; i < 3; i++){
        x_ci[i] = interact[i] - central[i];
        while (x_ci[i] > 0.5f * box_size[i]) x_ci[i] -= box_size[i];
        while (x_ci[i] < -0.5f * box_size[i]) x_ci[i] += box_size[i];
    }
}
